unit_tests/src/test_atomic_apply_acl_operations.cpp

```diff
@@ -512,6 +434,205 @@ TEST_CASE("#7338")
+                },
+                {
+                    {"entity_name", local_zone_user.name},
+                    {"acl", "null"}
+                },
+                {
+                    {"entity_name", remote_only_user.name},
+                    {"zone", test_zone_name},
+                    {"acl", "null"}
+                },
+            })}
+        }.dump();
+        // clang-format on
+
+        char* json_error_string{};
+        irods::at_scope_exit free_memory{[&json_error_string] { std::free(json_error_string); }};
+
+        REQUIRE(rc_atomic_apply_acl_operations(static_cast<rcComm_t*>(conn), json_input.c_str(), &json_error_string) ==
+                0);
+        REQUIRE(json_error_string == "{}"s);
+
+        const auto perms = fs::client::status(conn, user_home).permissions();
+        auto b = std::begin(perms);
+        auto e = std::end(perms);
+
+        for (auto&& u : {remote_zone_user, local_zone_user, remote_only_user}) {
+            REQUIRE(std::find_if(b, e, [&u](const fs::entity_permission& p) { return p.name == u.name; }) == e);
+        }
+    }};
+
+    // clang-format off
+    const auto failing_json_input = json{
+        {"logical_path", user_home},
+        {"operations", json::array({
+            {
+                {"entity_name", remote_only_user_name},
+                {"acl", "own"}
+            }
+        })}
+    }.dump();
+
+    const auto success_json_input = json{
+        {"logical_path", user_home},
+        {"operations", json::array({
+            {
+                {"entity_name", acl_user_name},
+                {"zone", test_zone_name},
+                {"acl", "write"}
+            }
+        })}
+    }.dump();
+
+    const auto remote_only_success_json_input = json{
+        {"logical_path", user_home},
+        {"operations", json::array({
+            {
+                {"entity_name", remote_only_user_name},
+                {"zone", test_zone_name},
+                {"acl", "own"}
+            }
+        })}
+    }.dump();
+
+    const auto local_success_json_input = json{
+        {"logical_path", user_home},
+        {"operations", json::array({
+            {
+                {"entity_name", acl_user_name},
+                {"acl", "read"}
+            }
+        })}
+    }.dump();
+
+    // clang-format off
+
+    char* json_error_string{};
+
+    // should fail: remote zone only user, but no zone specified
+    REQUIRE(rc_atomic_apply_acl_operations(
+                static_cast<rcComm_t*>(conn), failing_json_input.c_str(), &json_error_string) == SYS_INVALID_INPUT_PARAM);
+    REQUIRE(json_error_string != "{}"s);
+    std::free(json_error_string);
+
+    // should succeed: remote zone only user with zone param
+    REQUIRE(rc_atomic_apply_acl_operations(
+                static_cast<rcComm_t*>(conn), remote_only_success_json_input.c_str(), &json_error_string) == 0);
+    REQUIRE(json_error_string == "{}"s);
+    std::free(json_error_string);
```

Comment:
Perhaps it doesn't matter because this is a test, but... if the `REQUIRE` on the assertion on the above line fails, `json_error_string` will leak, right?

Reply:
I think what I did just now should take care of that. Could you double-check my work?

Reply:
I don't remember exactly what situation causes this, but pushing addresses onto `std::vector`s can lead to problems if the `std::vector` reallocates. I think because the `std::vector` is not taking ownership of the memory, this is okay, but it may be in our best interests to use `std::vector::reserve` just in case. I wish I could explain this better lol

Reply:
Alternatively, we could use a `std::array<char*, 4>` since we know the number of items. But like I said, I think what you have should be safe/okay.